In [28]:
cd MacroLLM

/content/drive/MyDrive/MacroLLM


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
cd drive/MyDrive/MacroLLM/

/content/drive/MyDrive/MacroLLM


In [23]:
ls

Construction_modele_v2_DistilRoBERTa.ipynb  Forex_data_corrected.csv
dataset_with_embeddings_distilroberta.pkl   logs/
distilroberta_forex_final/                  wandb/
distilroberta_forex_finetuned/


In [ ]:
# Installation des dépendances (décommentez si nécessaire)
# !pip install transformers torch datasets scikit-learn pandas numpy tqdm accelerate

In [ ]:
import pandas as pd
import numpy as np
import torch
from torch.utils.data import DataLoader, Dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    AutoModel,
    TrainingArguments,
    Trainer
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, classification_report
from tqdm import tqdm
import warnings
warnings.filterwarnings('ignore')

# Vérification GPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"🖥️ Device utilisé : {device}")
if torch.cuda.is_available():
    print(f"   GPU : {torch.cuda.get_device_name(0)}")

🖥️ Device utilisé : cuda
   GPU : Tesla T4


## 1. Chargement et Préparation des Données

In [ ]:
# Chargement du dataset CORRIGÉ (avec Embedding_Context et Label déjà créés)
data = pd.read_csv('Forex_data_corrected.csv')

print(f"📊 Dataset chargé : {len(data)} lignes")
print(f"Colonnes : {list(data.columns)}")
data.head(3)

📊 Dataset chargé : 15490 lignes
Colonnes : ['DateTime', 'Currency', 'Impact', 'Event', 'Actual', 'Forecast', 'Previous', 'Detail', 'Usual_Effect', 'Embedding_Context', 'Price_Variation', 'Label']


DateTime Currency                  Impact  \
0  2007-01-02 09:00:00+00:00      EUR  Medium Impact Expected   
1  2007-01-02 09:30:00+00:00      GBP  Medium Impact Expected   
2  2007-01-03 13:15:00+00:00      USD  Medium Impact Expected   

                            Event Actual Forecast Previous  \
0         Final Manufacturing PMI   56.5     56.5     56.6   
1         Final Manufacturing PMI   51.9     53.0     52.5   
2  ADP Non-Farm Employment Change   -40K     120K     230K   

                                              Detail  \
0  Source: S&P Global (latest release) | Measures...   
1  Source: S&P Global (latest release) | Measures...   
2  Source: Automatic Data Processing, Inc. (lates...   

                                        Usual_Effect  \
0  Actual greater than Forecast is good for currency   
1  Actual greater than Forecast is good for currency   
2  Actual greater than Forecast is good for currency   

                                   Embedding_Context  Price_Variation  Label  
0  EUR Final Manufacturing PMI. Actual: 56.5, For...         0.004315      2  
1  GBP Final Manufacturing PMI. Actual: 51.9, For...         0.004736      2  
2  USD ADP Non-Farm Employment Change. Actual: -4...         0.007903      2

In [ ]:
# Vérification des colonnes clés : Embedding_Context, Price_Variation, Label
print("📋 Vérification des colonnes clés :")
print(f"\n1. Embedding_Context (texte pour le fine-tuning) :")
print(f"   Exemple : {data['Embedding_Context'].iloc[0]}")

print(f"\n2. Price_Variation :")
print(data['Price_Variation'].describe())

print(f"\n3. Label (déjà créé avec seuil ±0.00005) :")
print(data['Label'].value_counts().sort_index())
print(f"\n   Valeurs manquantes : {data['Label'].isna().sum()}")

📋 Vérification des colonnes clés :

1. Embedding_Context (texte pour le fine-tuning) :
   Exemple : EUR Final Manufacturing PMI. Actual: 56.5, Forecast: 56.5. Usual Effect: Actual greater than Forecast is good for currency.

2. Price_Variation :
count    15490.000000
mean        -0.000034
std          0.003627
min         -0.056817
25%         -0.000093
50%         -0.000000
75%          0.000099
max          0.076135
Name: Price_Variation, dtype: float64

3. Label (déjà créé avec seuil ±0.00005) :
Label
0    4822
1    5601
2    5067
Name: count, dtype: int64

   Valeurs manquantes : 0


## 2. Préparation des Données pour le Fine-Tuning

Les labels ont déjà été créés dans le dataset `Forex_data_corrected.csv` avec le seuil optimal :
- **0 = BAISSE** : Price_Variation < -0.00005
- **1 = NEUTRE** : -0.00005 <= Price_Variation <= +0.00005  
- **2 = HAUSSE** : Price_Variation > +0.00005

La colonne `Embedding_Context` contient le texte formaté :
```
{Currency} {Event}. Actual: {Actual}, Forecast: {Forecast}. Usual Effect: {Usual Effect}.
```

In [ ]:
# Préparation du dataset pour le fine-tuning
# On utilise directement les colonnes Embedding_Context et Label du dataset

# Filtrer les lignes avec Label valide (non NaN)
df_labeled = data[data['Label'].notna()].copy()
df_labeled['Label'] = df_labeled['Label'].astype(int)

print(f"📊 Dataset pour fine-tuning : {len(df_labeled)} lignes")
print(f"   Seuil utilisé : ±0.00005 (±0.005%)")
print(f"\n🏷️ Distribution des labels (Classification Ternaire) :")
print(df_labeled['Label'].value_counts().sort_index())
print(f"\n   0 (BAISSE) : {(df_labeled['Label'] == 0).sum()} ({(df_labeled['Label'] == 0).mean():.1%})")
print(f"   1 (NEUTRE) : {(df_labeled['Label'] == 1).sum()} ({(df_labeled['Label'] == 1).mean():.1%})")
print(f"   2 (HAUSSE) : {(df_labeled['Label'] == 2).sum()} ({(df_labeled['Label'] == 2).mean():.1%})")

print(f"\n📝 Exemple de Embedding_Context :")
print(f"   {df_labeled['Embedding_Context'].iloc[0]}")

📊 Dataset pour fine-tuning : 15490 lignes
   Seuil utilisé : ±0.00005 (±0.005%)

🏷️ Distribution des labels (Classification Ternaire) :
Label
0    4822
1    5601
2    5067
Name: count, dtype: int64

   0 (BAISSE) : 4822 (31.1%)
   1 (NEUTRE) : 5601 (36.2%)
   2 (HAUSSE) : 5067 (32.7%)

📝 Exemple de Embedding_Context :
   EUR Final Manufacturing PMI. Actual: 56.5, Forecast: 56.5. Usual Effect: Actual greater than Forecast is good for currency.


In [ ]:
# Split Train / Validation / Test
# ⚠️ On utilise un split STRATIFIÉ pour garantir la même distribution de classes

from sklearn.model_selection import train_test_split

df_labeled['DateTime_parsed'] = pd.to_datetime(df_labeled['DateTime'], errors='coerce')

# Split stratifié : 70% Train, 15% Validation, 15% Test
# Étape 1 : Train (70%) vs Temp (30%)
train_df, temp_df = train_test_split(
    df_labeled,
    test_size=0.30,
    stratify=df_labeled['Label'],  # <-- Garantit l'équilibre des classes
    random_state=42
)

# Étape 2 : Validation (15%) vs Test (15%) - soit 50/50 du temp
val_df, test_df = train_test_split(
    temp_df,
    test_size=0.50,
    stratify=temp_df['Label'],  # <-- Garantit l'équilibre des classes
    random_state=42
)

print(f"📂 Split des données (STRATIFIÉ - équilibre garanti) :")
print(f"   Train      : {len(train_df)} lignes")
print(f"   Validation : {len(val_df)} lignes")
print(f"   Test       : {len(test_df)} lignes")

print(f"\n🏷️ Distribution des labels par split :")
for name, df in [("Train", train_df), ("Val", val_df), ("Test", test_df)]:
    print(f"\n   {name}:")
    for label in [0, 1, 2]:
        pct = (df['Label'] == label).mean()
        label_name = {0: "BAISSE", 1: "NEUTRE", 2: "HAUSSE"}[label]
        print(f"      {label} ({label_name}) : {pct:.1%}")

📂 Split des données (STRATIFIÉ - équilibre garanti) :
   Train      : 10843 lignes
   Validation : 2323 lignes
   Test       : 2324 lignes

🏷️ Distribution des labels par split :

   Train:
      0 (BAISSE) : 31.1%
      1 (NEUTRE) : 36.2%
      2 (HAUSSE) : 32.7%

   Val:
      0 (BAISSE) : 31.1%
      1 (NEUTRE) : 36.2%
      2 (HAUSSE) : 32.7%

   Test:
      0 (BAISSE) : 31.2%
      1 (NEUTRE) : 36.1%
      2 (HAUSSE) : 32.7%


## 3. Chargement de DistilRoBERTa

In [ ]:
# Chargement du modèle et tokenizer DistilRoBERTa
MODEL_NAME = "distilroberta-base"

print(f"🤖 Chargement de {MODEL_NAME}...")
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

# Modèle pour classification (3 classes : Baisse/Neutre/Hausse)
model = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME,
    num_labels=3,  # <-- 3 classes maintenant
    id2label={0: "BAISSE", 1: "NEUTRE", 2: "HAUSSE"},
    label2id={"BAISSE": 0, "NEUTRE": 1, "HAUSSE": 2}
)

model = model.to(device)
print(f"✅ Modèle chargé sur {device}")
print(f"   Paramètres : {model.num_parameters():,}")
print(f"   Classes : BAISSE (0), NEUTRE (1), HAUSSE (2)")

🤖 Chargement de distilroberta-base...


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at distilroberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


✅ Modèle chargé sur cuda
   Paramètres : 82,120,707
   Classes : BAISSE (0), NEUTRE (1), HAUSSE (2)


## 4. Préparation du Dataset pour Hugging Face Trainer

In [ ]:
from datasets import Dataset as HFDataset

# Conversion en format Hugging Face Dataset
# On utilise Embedding_Context au lieu de Text_Combined
def prepare_hf_dataset(df, tokenizer, max_length=256):
    """
    Prépare un DataFrame pour l'entraînement avec HuggingFace.
    Utilise la colonne Embedding_Context pour le texte.
    """
    # Créer le dataset HF avec Embedding_Context
    hf_dataset = HFDataset.from_pandas(df[['Embedding_Context', 'Label']].reset_index(drop=True))

    # Fonction de tokenisation
    def tokenize_function(examples):
        return tokenizer(
            examples['Embedding_Context'],  # <-- Utilise Embedding_Context
            padding='max_length',
            truncation=True,
            max_length=max_length
        )

    # Appliquer la tokenisation
    tokenized_dataset = hf_dataset.map(tokenize_function, batched=True)

    # Renommer la colonne Label en labels (requis par Trainer)
    tokenized_dataset = tokenized_dataset.rename_column('Label', 'labels')

    # Définir le format PyTorch
    tokenized_dataset.set_format('torch', columns=['input_ids', 'attention_mask', 'labels'])

    return tokenized_dataset

print("🔄 Tokenisation des datasets...")
train_dataset = prepare_hf_dataset(train_df, tokenizer)
val_dataset = prepare_hf_dataset(val_df, tokenizer)
test_dataset = prepare_hf_dataset(test_df, tokenizer)

print(f"✅ Datasets préparés (avec Embedding_Context) :")
print(f"   Train : {len(train_dataset)} exemples")
print(f"   Val   : {len(val_dataset)} exemples")
print(f"   Test  : {len(test_dataset)} exemples")

🔄 Tokenisation des datasets...


Map:   0%|          | 0/10843 [00:00<?, ? examples/s]

Map:   0%|          | 0/2323 [00:00<?, ? examples/s]

Map:   0%|          | 0/2324 [00:00<?, ? examples/s]

✅ Datasets préparés (avec Embedding_Context) :
   Train : 10843 exemples
   Val   : 2323 exemples
   Test  : 2324 exemples


## 5. Configuration et Lancement du Fine-Tuning

In [ ]:
# Fonction de calcul des métriques (adaptée pour 3 classes)
def compute_metrics(eval_pred):
    """
    Calcule les métriques d'évaluation pendant l'entraînement.
    Adapté pour la classification ternaire (3 classes).
    """
    predictions, labels = eval_pred
    predictions = np.argmax(predictions, axis=1)

    # Métriques macro (moyenne sur toutes les classes)
    precision, recall, f1, _ = precision_recall_fscore_support(
        labels, predictions, average='macro'  # 'macro' pour 3 classes
    )
    accuracy = accuracy_score(labels, predictions)

    return {
        'accuracy': accuracy,
        'precision': precision,
        'recall': recall,
        'f1': f1
    }

print("✅ Fonction de métriques définie (mode ternaire)")

✅ Fonction de métriques définie (mode ternaire)


In [ ]:
# Configuration de l'entraînement
training_args = TrainingArguments(
    output_dir='./distilroberta_forex_finetuned',

    # Hyperparamètres
    num_train_epochs=5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=32,
    learning_rate=2e-5,
    weight_decay=0.01,
    warmup_ratio=0.1,

    # Évaluation
    eval_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    metric_for_best_model="f1",
    greater_is_better=True,

    # Logging
    logging_dir='./logs',
    logging_steps=100,

    # Optimisation
    fp16=torch.cuda.is_available(),  # Mixed precision si GPU

    # Reproductibilité
    seed=42,
)

print("✅ Configuration de l'entraînement :")
print(f"   Epochs : {training_args.num_train_epochs}")
print(f"   Batch size : {training_args.per_device_train_batch_size}")
print(f"   Learning rate : {training_args.learning_rate}")
print(f"   FP16 : {training_args.fp16}")

✅ Configuration de l'entraînement :
   Epochs : 5
   Batch size : 16
   Learning rate : 2e-05
   FP16 : True


In [ ]:
# Création du Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics,
)

print("✅ Trainer configuré, prêt pour l'entraînement")

✅ Trainer configuré, prêt pour l'entraînement


In [ ]:
# 🚀 LANCEMENT DU FINE-TUNING
# ⚠️ Cette cellule peut prendre 10-30 minutes selon votre GPU/CPU

print("=" * 60)
print("🚀 DÉBUT DU FINE-TUNING DE DISTILROBERTA")
print("=" * 60)

train_result = trainer.train()

print("\n" + "=" * 60)
print("✅ FINE-TUNING TERMINÉ !")
print("=" * 60)
print(f"\nTemps d'entraînement : {train_result.metrics['train_runtime']:.1f} secondes")
print(f"Loss finale : {train_result.metrics['train_loss']:.4f}")

🚀 DÉBUT DU FINE-TUNING DE DISTILROBERTA



✅ FINE-TUNING TERMINÉ !

Temps d'entraînement : 406.4 secondes
Loss finale : 1.0893


In [ ]:
print("Distribution des labels par split :")
for name, df in [("Train", train_df), ("Val", val_df), ("Test", test_df)]:
    print(f"\n{name}:")
    print(df['Label'].value_counts(normalize=True).sort_index())

Distribution des labels par split :

Train:
Label
0    0.311261
1    0.361616
2    0.327123
Name: proportion, dtype: float64

Val:
Label
0    0.311235
1    0.361601
2    0.327163
Name: proportion, dtype: float64

Test:
Label
0    0.311532
1    0.361446
2    0.327022
Name: proportion, dtype: float64


In [ ]:
# Sauvegarde du modèle fine-tuné
MODEL_SAVE_PATH = './distilroberta_forex_final'

trainer.save_model(MODEL_SAVE_PATH)
tokenizer.save_pretrained(MODEL_SAVE_PATH)

print(f"💾 Modèle sauvegardé dans : {MODEL_SAVE_PATH}")

💾 Modèle sauvegardé dans : ./distilroberta_forex_final


## 6. Évaluation sur le Set de Test

In [ ]:
# Évaluation finale sur le set de test (3 classes)
print("=" * 60)
print("📊 ÉVALUATION SUR LE SET DE TEST (Classification Ternaire)")
print("=" * 60)

test_results = trainer.evaluate(test_dataset)

print(f"""
┌─────────────────────────────────────────────────────────────────────┐
│        MÉTRIQUES DU MODÈLE FINE-TUNÉ (TEST SET - 3 CLASSES)        │
├─────────────────────────────────────────────────────────────────────┤
│  Accuracy (Exactitude)     : {test_results['eval_accuracy']:.2%}                              │
│  Precision (Macro)         : {test_results['eval_precision']:.2%}                              │
│  Recall (Macro)            : {test_results['eval_recall']:.2%}                              │
│  F1-Score (Macro)          : {test_results['eval_f1']:.2%}                              │
└─────────────────────────────────────────────────────────────────────┘
""")

📊 ÉVALUATION SUR LE SET DE TEST (Classification Ternaire)



┌─────────────────────────────────────────────────────────────────────┐
│        MÉTRIQUES DU MODÈLE FINE-TUNÉ (TEST SET - 3 CLASSES)        │
├─────────────────────────────────────────────────────────────────────┤
│  Accuracy (Exactitude)     : 38.51%                              │
│  Precision (Macro)         : 37.13%                              │
│  Recall (Macro)            : 37.22%                              │
│  F1-Score (Macro)          : 33.93%                              │
└─────────────────────────────────────────────────────────────────────┘



In [ ]:
# Rapport de classification détaillé (3 classes)
from sklearn.metrics import confusion_matrix

# Prédictions sur le test set
predictions = trainer.predict(test_dataset)
y_pred = np.argmax(predictions.predictions, axis=1)
y_true = predictions.label_ids

print("📋 RAPPORT DE CLASSIFICATION DÉTAILLÉ (3 CLASSES) :")
print(classification_report(y_true, y_pred, target_names=['BAISSE', 'NEUTRE', 'HAUSSE']))

print("\n📊 MATRICE DE CONFUSION (3x3) :")
cm = confusion_matrix(y_true, y_pred)
print("                        Prédit")
print("                 BAISSE  NEUTRE  HAUSSE")
print(f"   Réel BAISSE    {cm[0][0]:>4}    {cm[0][1]:>4}    {cm[0][2]:>4}")
print(f"        NEUTRE    {cm[1][0]:>4}    {cm[1][1]:>4}    {cm[1][2]:>4}")
print(f"        HAUSSE    {cm[2][0]:>4}    {cm[2][1]:>4}    {cm[2][2]:>4}")

📋 RAPPORT DE CLASSIFICATION DÉTAILLÉ (3 CLASSES) :
              precision    recall  f1-score   support

      BAISSE       0.34      0.09      0.14       724
      NEUTRE       0.40      0.62      0.49       840
      HAUSSE       0.37      0.41      0.39       760

    accuracy                           0.39      2324
   macro avg       0.37      0.37      0.34      2324
weighted avg       0.37      0.39      0.35      2324


📊 MATRICE DE CONFUSION (3x3) :
                        Prédit
                 BAISSE  NEUTRE  HAUSSE
   Réel BAISSE      66     398     260
        NEUTRE      50     521     269
        HAUSSE      76     376     308


## 7. Extraction des Embeddings avec le Modèle Fine-Tuné

In [ ]:
# Charger le modèle de base (sans la tête de classification) pour les embeddings
# On utilise le modèle fine-tuné mais on extrait les embeddings de la couche cachée

print("🔄 Chargement du modèle pour extraction d'embeddings...")

# On charge le modèle fine-tuné
model_for_embeddings = AutoModel.from_pretrained(MODEL_SAVE_PATH)
model_for_embeddings = model_for_embeddings.to(device)
model_for_embeddings.eval()

print(f"✅ Modèle chargé pour embeddings (dimension : 768)")

Some weights of RobertaModel were not initialized from the model checkpoint at ./distilroberta_forex_final and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


🔄 Chargement du modèle pour extraction d'embeddings...
✅ Modèle chargé pour embeddings (dimension : 768)


In [11]:
# Fonction d'extraction d'embeddings avec le modèle fine-tuné
def get_embeddings_distilroberta(texts, batch_size=32):
    """
    Extrait les embeddings avec notre DistilRoBERTa fine-tuné.
    Utilise Mean Pooling sur les tokens.
    """
    all_embeddings = []

    for i in tqdm(range(0, len(texts), batch_size), desc="Extraction embeddings"):
        batch_texts = texts[i:i + batch_size]

        # Tokenisation
        encoded = tokenizer(
            batch_texts,
            padding=True,
            truncation=True,
            max_length=256,
            return_tensors='pt'
        )

        encoded = {key: val.to(device) for key, val in encoded.items()}

        # Inférence
        with torch.no_grad():
            outputs = model_for_embeddings(**encoded)

        # Mean Pooling
        attention_mask = encoded['attention_mask']
        token_embeddings = outputs.last_hidden_state

        mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
        sum_embeddings = torch.sum(token_embeddings * mask_expanded, dim=1)
        sum_mask = torch.clamp(mask_expanded.sum(dim=1), min=1e-9)
        mean_embeddings = sum_embeddings / sum_mask

        all_embeddings.extend(mean_embeddings.cpu().numpy())

    return all_embeddings

print("✅ Fonction d'extraction d'embeddings prête")

✅ Fonction d'extraction d'embeddings prête


In [ ]:
# Application des embeddings sur tout le dataset
# On utilise Embedding_Context pour générer les embeddings
# ⚠️ Cette cellule peut prendre plusieurs minutes

print(f"🔄 Extraction des embeddings pour {len(data)} lignes...")

# Utiliser Embedding_Context au lieu de Text_Combined
texts_list = data['Embedding_Context'].tolist()
embeddings = get_embeddings_distilroberta(texts_list, batch_size=32)

data['Embedding'] = embeddings

print(f"\n✅ Embeddings extraits !")
print(f"   Dimension : {len(embeddings[0])} (768)")
print(f"   Exemple : {embeddings[0][:5]}")

🔄 Extraction des embeddings pour 15490 lignes...


Extraction embeddings: 100%|██████████| 485/485 [00:18<00:00, 26.75it/s]


✅ Embeddings extraits !
   Dimension : 768 (768)
   Exemple : [ 0.05902129  0.3082685   0.06456606 -0.20115116 -0.7098534 ]


In [ ]:
# Sauvegarde du dataset avec embeddings DistilRoBERTa
output_path = 'dataset_with_embeddings_distilroberta.pkl'
data.to_pickle(output_path)

print(f"💾 Dataset sauvegardé : {output_path}")
print(f"   Colonnes : {list(data.columns)}")

💾 Dataset sauvegardé : dataset_with_embeddings_distilroberta.pkl
   Colonnes : ['DateTime', 'Currency', 'Impact', 'Event', 'Actual', 'Forecast', 'Previous', 'Detail', 'Usual_Effect', 'Embedding_Context', 'Price_Variation', 'Label', 'Embedding']


In [5]:
ls

Construction_modele_v2_DistilRoBERTa.ipynb  Forex_data_corrected.csv
dataset_with_embeddings_distilroberta.pkl   logs/
distilroberta_forex_final/                  wandb/
distilroberta_forex_finetuned/


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
cd drive/MyDrive/MacroLLM/

/content/drive/MyDrive/MacroLLM


## ⚡ RECHARGEMENT RAPIDE (Si vous revenez après une déconnexion)

**Exécutez cette cellule si vous avez déjà entraîné et sauvegardé le modèle.**  
Elle recharge toutes les variables nécessaires pour la suite (similarité, sentiment, RAG, Gemini).

⚠️ **Prérequis** :
- Le modèle doit avoir été sauvegardé dans `./distilroberta_forex_final/`
- Le dataset avec embeddings doit exister : `dataset_with_embeddings_distilroberta.pkl`

In [6]:
# ═══════════════════════════════════════════════════════════════════════════════
# ⚡ CELLULE DE RECHARGEMENT RAPIDE
# ═══════════════════════════════════════════════════════════════════════════════
# Exécutez cette cellule UNIQUEMENT si vous revenez après une déconnexion
# et que le modèle a déjà été entraîné et sauvegardé.
#
# Cette cellule charge UNIQUEMENT les variables nécessaires.
# Les fonctions (get_sentiment, find_similar_events, etc.) sont définies
# dans les cellules suivantes - exécutez-les ensuite.
# ═══════════════════════════════════════════════════════════════════════════════

import pandas as pd
import numpy as np
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification, AutoModel
from sklearn.metrics.pairwise import cosine_similarity
from tqdm import tqdm
import os
import warnings
warnings.filterwarnings('ignore')

print("=" * 70)
print("⚡ RECHARGEMENT RAPIDE DU MODÈLE FINE-TUNÉ")
print("=" * 70)

# 1. Device (GPU/CPU)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"\n🖥️ Device : {device}")

# 2. Chemin du modèle
MODEL_SAVE_PATH = './distilroberta_forex_final'

# 3. Tokenizer
print("\n📦 Chargement du tokenizer...")
tokenizer = AutoTokenizer.from_pretrained(MODEL_SAVE_PATH)
print("   ✅ Tokenizer chargé")

# 4. Modèle de classification (pour le sentiment) - utilisé par get_sentiment_distilroberta()
print("\n🧠 Chargement du modèle de classification...")
model_classifier = AutoModelForSequenceClassification.from_pretrained(MODEL_SAVE_PATH)
model_classifier = model_classifier.to(device)
model_classifier.eval()
print("   ✅ model_classifier chargé")

# 5. Modèle pour embeddings (pour la similarité) - utilisé par get_embeddings_distilroberta()
print("\n🔢 Chargement du modèle pour embeddings...")
model_for_embeddings = AutoModel.from_pretrained(MODEL_SAVE_PATH)
model_for_embeddings = model_for_embeddings.to(device)
model_for_embeddings.eval()
print("   ✅ model_for_embeddings chargé")

# 6. Dataset avec embeddings pré-calculés
print("\n📊 Chargement du dataset...")
if os.path.exists('dataset_with_embeddings_distilroberta.pkl'):
    data = pd.read_pickle('dataset_with_embeddings_distilroberta.pkl')
    print(f"   ✅ Dataset chargé (pickle) : {len(data)} lignes avec embeddings")
else:
    data = pd.read_csv('Forex_data_corrected.csv')
    print(f"   ⚠️ Pickle non trouvé, CSV chargé : {len(data)} lignes")
    print("   ⚠️ Exécutez la Section 7 pour générer les embeddings")

# Résumé
print("\n" + "=" * 70)
print("✅ RECHARGEMENT TERMINÉ")
print("=" * 70)
print("""
📦 Variables chargées :
   - device, MODEL_SAVE_PATH, tokenizer
   - model_classifier, model_for_embeddings
   - data

🔜 Prochaine étape : Exécutez les cellules suivantes pour définir les fonctions :
   1. Cellule 28 : get_embeddings_distilroberta()  ← IMPORTANT (Section 7)
   2. Cellule 34 : get_sentiment_distilroberta()   (Section 8)
   3. Cellule 37 : find_similar_events_v2()        (Section 9)
   4. Cellule 40 : generate_analysis_v2()          (Section 11)
""")

⚡ RECHARGEMENT RAPIDE DU MODÈLE FINE-TUNÉ

🖥️ Device : cpu

📦 Chargement du tokenizer...
   ✅ Tokenizer chargé

🧠 Chargement du modèle de classification...


   ✅ model_classifier chargé

🔢 Chargement du modèle pour embeddings...


Some weights of RobertaModel were not initialized from the model checkpoint at ./distilroberta_forex_final and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


   ✅ model_for_embeddings chargé

📊 Chargement du dataset...
   ✅ Dataset chargé (pickle) : 15490 lignes avec embeddings

✅ RECHARGEMENT TERMINÉ

📦 Variables chargées :
   - device, MODEL_SAVE_PATH, tokenizer
   - model_classifier, model_for_embeddings
   - data

🔜 Prochaine étape : Exécutez les cellules suivantes pour définir les fonctions :
   - Cellule 34 : get_sentiment_distilroberta()
   - Cellule 35 : Test sur 5 exemples
   - Cellule 37 : find_similar_events_v2()
   - Cellule 40 : generate_analysis_v2()



In [ ]:
ls

config.json  special_tokens_map.json  tokenizer.json     vocab.json
merges.txt   tokenizer_config.json    training_args.bin


In [ ]:
cd distilroberta_forex_final/

/content/drive/MyDrive/MacroLLM/distilroberta_forex_final


## 8. Fonction de Prédiction de Sentiment (avec notre modèle)

In [7]:
# Charger le modèle de classification fine-tuné pour les prédictions (3 classes)
model_classifier = AutoModelForSequenceClassification.from_pretrained(MODEL_SAVE_PATH)
model_classifier = model_classifier.to(device)
model_classifier.eval()

def get_sentiment_distilroberta(text):
    """
    Prédit le sentiment (BAISSE/NEUTRE/HAUSSE) avec notre modèle fine-tuné.

    Args:
        text: Texte au format Embedding_Context
              Ex: "EUR Final Manufacturing PMI. Actual: 56.5, Forecast: 56.5. Usual Effect: ..."

    Returns:
        dict: {'label': 'negative/neutral/positive', 'score': 0.0-1.0, ...}
    """
    # Tokenisation
    encoded = tokenizer(
        text,
        padding=True,
        truncation=True,
        max_length=256,
        return_tensors='pt'
    )
    encoded = {key: val.to(device) for key, val in encoded.items()}

    # Prédiction
    with torch.no_grad():
        outputs = model_classifier(**encoded)

    # Softmax pour obtenir les probabilités
    probs = torch.nn.functional.softmax(outputs.logits, dim=-1)
    predicted_class = torch.argmax(probs, dim=-1).item()
    confidence = probs[0][predicted_class].item()

    # Mapping vers le format attendu (compatible avec le pipeline RAG)
    label_map = {0: "negative", 1: "neutral", 2: "positive"}
    label = label_map[predicted_class]

    # Détails pour affichage
    class_names = {0: "BAISSE", 1: "NEUTRE", 2: "HAUSSE"}

    return {
        'label': label,
        'score': confidence,
        'class_name': class_names[predicted_class],
        'all_probabilities': {
            'BAISSE': probs[0][0].item(),
            'NEUTRE': probs[0][1].item(),
            'HAUSSE': probs[0][2].item()
        }
    }

# Test avec le nouveau format Embedding_Context
test_text = "USD Non-Farm Employment Change. Actual: 250K, Forecast: 180K. Usual Effect: Actual greater than Forecast is good for currency."
result = get_sentiment_distilroberta(test_text)
print(f"🧪 Test de sentiment (Classification Ternaire) :")
print(f"   Texte : {test_text}")
print(f"   Prédiction : {result['class_name']} (confiance: {result['score']:.2%})")
print(f"   Probabilités : BAISSE={result['all_probabilities']['BAISSE']:.1%}, NEUTRE={result['all_probabilities']['NEUTRE']:.1%}, HAUSSE={result['all_probabilities']['HAUSSE']:.1%}")

🧪 Test de sentiment (Classification Ternaire) :
   Texte : USD Non-Farm Employment Change. Actual: 250K, Forecast: 180K. Usual Effect: Actual greater than Forecast is good for currency.
   Prédiction : NEUTRE (confiance: 40.76%)
   Probabilités : BAISSE=32.8%, NEUTRE=40.8%, HAUSSE=26.4%


In [8]:
# 5 exemples de test pour évaluer le modèle

test_examples = [
    # 1. CPI m/m - Inflation en hausse (devrait être HAUSSE pour USD)
    "USD CPI m/m. Actual: 0.4%, Forecast: 0.2%. Usual Effect: Actual greater than Forecast is good for currency.",

    # 2. Unemployment Rate - Chômage en baisse (devrait être HAUSSE pour EUR)
    "EUR Unemployment Rate. Actual: 6.4%, Forecast: 6.6%. Usual Effect: Actual less than Forecast is good for currency.",

    # 3. Federal Funds Rate - Taux maintenu (devrait être NEUTRE)
    "USD Federal Funds Rate. Actual: 5.50%, Forecast: 5.50%. Usual Effect: Actual greater than Forecast is good for currency.",

    # 4. Retail Sales m/m - Ventes en baisse (devrait être BAISSE pour GBP)
    "GBP Retail Sales m/m. Actual: -0.3%, Forecast: 0.2%. Usual Effect: Actual greater than Forecast is good for currency.",

    # 5. German ifo Business Climate - Confiance en hausse (devrait être HAUSSE pour EUR)
    "EUR German ifo Business Climate. Actual: 89.5, Forecast: 87.0. Usual Effect: Actual greater than Forecast is good for currency.",
]

# Test sur chaque exemple
print("=" * 70)
print("🧪 ÉVALUATION DU MODÈLE SUR 5 EXEMPLES DIVERSIFIÉS")
print("=" * 70)

for i, text in enumerate(test_examples, 1):
    result = get_sentiment_distilroberta(text)
    print(f"\n📌 Exemple {i}:")
    print(f"   Texte: {text[:80]}...")
    print(f"   Prédiction: {result['class_name']} (confiance: {result['score']:.1%})")
    print(f"   Probas: BAISSE={result['all_probabilities']['BAISSE']:.1%}, "
          f"NEUTRE={result['all_probabilities']['NEUTRE']:.1%}, "
          f"HAUSSE={result['all_probabilities']['HAUSSE']:.1%}")

🧪 ÉVALUATION DU MODÈLE SUR 5 EXEMPLES DIVERSIFIÉS

📌 Exemple 1:
   Texte: USD CPI m/m. Actual: 0.4%, Forecast: 0.2%. Usual Effect: Actual greater than For...
   Prédiction: NEUTRE (confiance: 41.0%)
   Probas: BAISSE=32.9%, NEUTRE=41.0%, HAUSSE=26.1%

📌 Exemple 2:
   Texte: EUR Unemployment Rate. Actual: 6.4%, Forecast: 6.6%. Usual Effect: Actual less t...
   Prédiction: NEUTRE (confiance: 41.3%)
   Probas: BAISSE=31.1%, NEUTRE=41.3%, HAUSSE=27.5%

📌 Exemple 3:
   Texte: USD Federal Funds Rate. Actual: 5.50%, Forecast: 5.50%. Usual Effect: Actual gre...
   Prédiction: NEUTRE (confiance: 40.0%)
   Probas: BAISSE=33.2%, NEUTRE=40.0%, HAUSSE=26.8%

📌 Exemple 4:
   Texte: GBP Retail Sales m/m. Actual: -0.3%, Forecast: 0.2%. Usual Effect: Actual greate...
   Prédiction: HAUSSE (confiance: 36.7%)
   Probas: BAISSE=36.7%, NEUTRE=26.5%, HAUSSE=36.7%

📌 Exemple 5:
   Texte: EUR German ifo Business Climate. Actual: 89.5, Forecast: 87.0. Usual Effect: Act...
   Prédiction: NEUTRE (confiance: 41.8

## 9. Pipeline RAG Complet (Version DistilRoBERTa)

In [9]:
# Fonctions RAG (Version DistilRoBERTa avec Embedding_Context)
from sklearn.metrics.pairwise import cosine_similarity
from datetime import datetime

def find_similar_events_v2(query_text, data, top_k=5, query_date=None):
    """
    Trouve les K événements historiques les plus similaires (Version DistilRoBERTa).

    Args:
        query_text: Texte au format Embedding_Context
        data: DataFrame avec la colonne Embedding
        top_k: Nombre de résultats
        query_date: Date de référence pour filtrage temporel
    """
    # Gestion de la date
    if query_date is None:
        reference_date = pd.Timestamp.now()
    elif isinstance(query_date, str):
        reference_date = pd.to_datetime(query_date)
    else:
        reference_date = pd.to_datetime(query_date)

    # Filtrage temporel
    data_filtered = data.copy()
    data_filtered['DateTime_parsed'] = pd.to_datetime(data_filtered['DateTime'], errors='coerce')

    # Supprimer le timezone si présent pour éviter l'erreur de comparaison
    if data_filtered['DateTime_parsed'].dt.tz is not None:
        data_filtered['DateTime_parsed'] = data_filtered['DateTime_parsed'].dt.tz_localize(None)
    if hasattr(reference_date, 'tz') and reference_date.tz is not None:
        reference_date = reference_date.tz_localize(None)

    data_filtered = data_filtered[data_filtered['DateTime_parsed'] < reference_date]

    if len(data_filtered) == 0:
        print(f"⚠️ Aucun événement trouvé avant {reference_date}")
        return pd.DataFrame()

    print(f"📅 Filtrage : {len(data_filtered)} événements avant {reference_date.strftime('%Y-%m-%d')}")

    # Vectorisation avec notre modèle
    query_embedding = get_embeddings_distilroberta([query_text], batch_size=1)[0]

    # Similarité cosinus
    embeddings_matrix = np.vstack(data_filtered['Embedding'].values)
    similarities = cosine_similarity([query_embedding], embeddings_matrix)[0]

    top_indices = similarities.argsort()[-top_k:][::-1]

    results = data_filtered.iloc[top_indices].copy()
    results['Similarity_Score'] = similarities[top_indices]

    return results[['DateTime', 'Currency', 'Impact', 'Event', 'Actual', 'Forecast', 'Previous', 'Price_Variation', 'Label', 'Similarity_Score']]

print("✅ Fonction de recherche de similarité v2 prête (avec Label)")

✅ Fonction de recherche de similarité v2 prête (avec Label)


In [12]:

# Test complet du pipeline v2 (avec classification ternaire et nouveau format)
# Format Embedding_Context : {Currency} {Event}. Actual: {Actual}, Forecast: {Forecast}. Usual Effect: {Usual Effect}.

#test_query = "CAD Unemployment Rate. Actual: 6.9%, Forecast: 7.1%. Usual Effect: Actual less than Forecast is good for currency."
test_query = "GBP Retail Sales m/m. Actual: -0.3%, Forecast: 0.2%. Usual Effect: Actual greater than Forecast is good for currency."
news_date = "2024-06-01"

print("=" * 60)
print("🚀 PIPELINE V2 - DISTILROBERTA FINE-TUNÉ (3 CLASSES)")
print("=" * 60)
print(f"\n📝 Format d'entrée : Embedding_Context")
print(f"   Query : {test_query}")

# Étape 1 : Recherche de similarité
print("\n📊 ÉTAPE 1 : Recherche de précédents...")
similar = find_similar_events_v2(test_query, data, top_k=5, query_date=news_date)
print(similar.to_string())

# Étape 2 : Sentiment avec notre modèle (3 classes)
print("\n🧠 ÉTAPE 2 : Analyse de sentiment (DistilRoBERTa - 3 classes)...")
sentiment = get_sentiment_distilroberta(test_query)
print(f"   Prédiction : {sentiment['class_name']} (confiance: {sentiment['score']:.1%})")
print(f"   Probabilités détaillées :")
print(f"      - BAISSE : {sentiment['all_probabilities']['BAISSE']:.1%}")
print(f"      - NEUTRE : {sentiment['all_probabilities']['NEUTRE']:.1%}")
print(f"      - HAUSSE : {sentiment['all_probabilities']['HAUSSE']:.1%}")

🚀 PIPELINE V2 - DISTILROBERTA FINE-TUNÉ (3 CLASSES)

📝 Format d'entrée : Embedding_Context
   Query : GBP Retail Sales m/m. Actual: -0.3%, Forecast: 0.2%. Usual Effect: Actual greater than Forecast is good for currency.

📊 ÉTAPE 1 : Recherche de précédents...
📅 Filtrage : 14750 événements avant 2024-06-01


Extraction embeddings: 100%|██████████| 1/1 [00:00<00:00,  6.50it/s]


                        DateTime Currency                  Impact             Event Actual Forecast Previous  Price_Variation  Label  Similarity_Score
11138  2020-03-25 19:30:00+00:00      GBP  Medium Impact Expected  Retail Sales m/m  -0.3%     0.2%     1.1%         0.000047      1          1.000000
13939  2023-06-23 05:00:00+00:00      GBP  Medium Impact Expected  Retail Sales m/m   0.3%    -0.2%     0.5%         0.000153      2          0.999969
9866   2018-09-20 08:30:00+00:00      GBP    High Impact Expected  Retail Sales m/m   0.3%    -0.2%     0.9%         0.000039      1          0.999969
13214  2022-08-18 19:30:00+00:00      GBP  Medium Impact Expected  Retail Sales m/m   0.3%    -0.2%    -0.2%        -0.000060      0          0.999969
10632  2019-08-15 08:30:00+00:00      GBP    High Impact Expected  Retail Sales m/m   0.2%    -0.3%     0.9%         0.000000      1          0.999950

🧠 ÉTAPE 2 : Analyse de sentiment (DistilRoBERTa - 3 classes)...
   Prédiction : HAUSSE (confi

11. Génération du Prompt pour le LLM (Gemini)

In [13]:
def generate_analysis_v2(news_text, similar_events_df, sentiment_result):
    """
    Génère un prompt d'analyse macro-économique complète pour un LLM.
    VERSION 2 : Utilise DistilRoBERTa fine-tuné avec classification ternaire.

    Args:
        news_text: La nouvelle news au format Embedding_Context
        similar_events_df: DataFrame des événements historiques similaires
        sentiment_result: Résultat de l'analyse de sentiment DistilRoBERTa (3 classes)

    Returns:
        str: Le prompt formaté pour le LLM
    """
    # Construction du contexte historique avec les labels
    historical_context = ""
    for idx, row in similar_events_df.iterrows():
        # Direction basée sur Price_Variation
        if row['Price_Variation'] > 0.00005:
            direction = "HAUSSE"
        elif row['Price_Variation'] < -0.00005:
            direction = "BAISSE"
        else:
            direction = "NEUTRE"

        variation = abs(row['Price_Variation'] * 100)

        # Gestion du format de date
        date_value = row['DateTime']
        if hasattr(date_value, 'strftime'):
            date_str = date_value.strftime('%Y-%m-%d')
        else:
            date_str = str(date_value)[:10]

        # Label historique
        label_map = {0: "BAISSE", 1: "NEUTRE", 2: "HAUSSE"}
        label_hist = label_map.get(row.get('Label', 1), "N/A")

        historical_context += f"""
    - {date_str} : {row['Event']} ({row['Currency']})
      Actual: {row['Actual']}, Forecast: {row['Forecast']}, Previous: {row['Previous']}
      Réaction marché : {direction} ({variation:.4f}%) - Label: {label_hist}
      Score de similarité : {row['Similarity_Score']:.4f}
"""

    # Interprétation du sentiment DistilRoBERTa (3 classes)
    class_name = sentiment_result['class_name']
    confidence = sentiment_result['score']
    probs = sentiment_result['all_probabilities']

    if class_name == "HAUSSE":
        sentiment_interpretation = "HAUSSIER pour la devise (signal d'achat potentiel)"
    elif class_name == "BAISSE":
        sentiment_interpretation = "BAISSIER pour la devise (signal de vente potentiel)"
    else:
        sentiment_interpretation = "NEUTRE (pas d'impact significatif attendu)"

    # Construction du prompt pour le LLM
    prompt = f"""Tu es un analyste macro-économique expert spécialisé dans le marché des devises (Forex).
Un trader te demande d'analyser l'annonce économique suivante et de lui donner des conseils.

=== NOUVELLE ANNONCE ===
{news_text}

=== ANALYSE DE SENTIMENT (DistilRoBERTa Fine-Tuné sur Forex) ===
Prédiction : {sentiment_interpretation}
Classe prédite : {class_name}
Niveau de confiance : {confidence:.1%}

Probabilités détaillées :
- BAISSE : {probs['BAISSE']:.1%}
- NEUTRE : {probs['NEUTRE']:.1%}
- HAUSSE : {probs['HAUSSE']:.1%}

Note : Ce modèle a été fine-tuné sur des données Forex réelles avec des labels basés sur les vraies réactions du marché (Price_Variation).

=== PRÉCÉDENTS HISTORIQUES SIMILAIRES (trouvés par DistilRoBERTa Embeddings) ===
{historical_context}

=== TA MISSION ===
En te basant sur l'analyse de sentiment ET les précédents historiques, génère une analyse complète qui inclut :

1. **EXPLICATION** : Explique simplement ce que signifie cette annonce économique.

2. **ANALYSE DE LA SURPRISE** : Compare Actual vs Forecast. Est-ce une bonne ou mauvaise surprise pour la devise ?

3. **VALIDATION DU SENTIMENT** : La prédiction du modèle ({class_name}) est-elle cohérente avec les chiffres annoncés ?

4. **ANALYSE HISTORIQUE** : Que s'est-il passé dans le passé avec des annonces similaires ? Quel pattern observes-tu ?

5. **SCÉNARIOS PROBABLES** :
   - Scénario Haussier (pour la devise) : conditions et probabilité
   - Scénario Baissier (pour la devise) : conditions et probabilité

6. **PAIRES À SURVEILLER** : Quelles paires de devises surveiller en priorité ?

7. **RECOMMANDATION** : Conseil pratique pour le trader (timing, prudence, confirmation à attendre, etc.)

Réponds de manière structurée et professionnelle.
"""

    return prompt

print("✅ Fonction generate_analysis_v2() prête (adaptée pour DistilRoBERTa 3 classes)")

✅ Fonction generate_analysis_v2() prête (adaptée pour DistilRoBERTa 3 classes)


## 12. Test Complet du Pipeline (Embeddings + Sentiment + Prompt)

In [14]:
# Test complet du pipeline V2 : Embeddings + Sentiment + Génération de Prompt
# Format Embedding_Context : {Currency} {Event}. Actual: {Actual}, Forecast: {Forecast}. Usual Effect: {Usual Effect}.

test_query = "CAD Unemployment Rate. Actual: 6.9%, Forecast: 7.1%. Usual Effect: Actual less than Forecast is good for currency."
news_date = "2024-06-01"

print("=" * 70)
print("🚀 PIPELINE COMPLET V2 - DISTILROBERTA FINE-TUNÉ")
print("=" * 70)
print(f"\n📝 Annonce à analyser :")
print(f"   {test_query}")
print(f"   Date de référence : {news_date}")

# ═══════════════════════════════════════════════════════════════════════
# ÉTAPE 1 : Recherche de précédents historiques (Embeddings DistilRoBERTa)
# ═══════════════════════════════════════════════════════════════════════
print("\n" + "─" * 70)
print("📊 ÉTAPE 1 : Recherche de précédents historiques")
print("─" * 70)
similar_events = find_similar_events_v2(test_query, data, top_k=5, query_date=news_date)
print("\n🔍 Top 5 événements similaires :")
print(similar_events.to_string())

# ═══════════════════════════════════════════════════════════════════════
# ÉTAPE 2 : Analyse de Sentiment (Classification DistilRoBERTa 3 classes)
# ═══════════════════════════════════════════════════════════════════════
print("\n" + "─" * 70)
print("🧠 ÉTAPE 2 : Analyse de sentiment (DistilRoBERTa - 3 classes)")
print("─" * 70)
sentiment = get_sentiment_distilroberta(test_query)
print(f"\n   Prédiction : {sentiment['class_name']} (confiance: {sentiment['score']:.1%})")
print(f"\n   Probabilités détaillées :")
print(f"      - BAISSE : {sentiment['all_probabilities']['BAISSE']:.1%}")
print(f"      - NEUTRE : {sentiment['all_probabilities']['NEUTRE']:.1%}")
print(f"      - HAUSSE : {sentiment['all_probabilities']['HAUSSE']:.1%}")

# ═══════════════════════════════════════════════════════════════════════
# ÉTAPE 3 : Génération du Prompt pour le LLM
# ═══════════════════════════════════════════════════════════════════════
print("\n" + "─" * 70)
print("📝 ÉTAPE 3 : Génération du prompt pour Gemini")
print("─" * 70)
prompt = generate_analysis_v2(test_query, similar_events, sentiment)

print("\n✅ Prompt généré avec succès !")
print(f"   Longueur : {len(prompt)} caractères")

# Affichage du prompt complet
print("\n" + "=" * 70)
print("📄 PROMPT FINAL POUR GEMINI")
print("=" * 70)
print(prompt)

🚀 PIPELINE COMPLET V2 - DISTILROBERTA FINE-TUNÉ

📝 Annonce à analyser :
   CAD Unemployment Rate. Actual: 6.9%, Forecast: 7.1%. Usual Effect: Actual less than Forecast is good for currency.
   Date de référence : 2024-06-01

──────────────────────────────────────────────────────────────────────
📊 ÉTAPE 1 : Recherche de précédents historiques
──────────────────────────────────────────────────────────────────────
📅 Filtrage : 14750 événements avant 2024-06-01


Extraction embeddings: 100%|██████████| 1/1 [00:00<00:00,  5.09it/s]



🔍 Top 5 événements similaires :
                       DateTime Currency                Impact              Event Actual Forecast Previous  Price_Variation  Label  Similarity_Score
7925  2016-06-09 19:30:00+00:00      CAD  High Impact Expected  Unemployment Rate   6.9%     7.1%     7.1%         0.000150      2          1.000000
5617  2013-10-10 20:30:00+00:00      CAD  High Impact Expected  Unemployment Rate   6.9%     7.1%     7.1%        -0.000077      0          1.000000
7341  2015-10-08 20:30:00+00:00      CAD  High Impact Expected  Unemployment Rate   7.1%     6.9%     7.0%        -0.000084      0          0.999956
5684  2013-11-07 20:30:00+00:00      CAD  High Impact Expected  Unemployment Rate   6.9%     7.0%     6.9%         0.000010      1          0.999813
6032  2014-04-03 19:30:00+00:00      CAD  High Impact Expected  Unemployment Rate   6.9%     7.0%     7.0%         0.000091      2          0.999813

──────────────────────────────────────────────────────────────────────
🧠

## 13. Appel au LLM Gemini (Google AI)

In [18]:
# !pip install google-generativeai

import google.generativeai as genai

# Configuration de la clé API Gemini
# Remplacez par votre clé API (obtenue sur https://makersuite.google.com/app/apikey)
GEMINI_API_KEY = "AIzaSyBExigWwXDNDc_J9X2wtQwvLimfIfX_jYA"
genai.configure(api_key=GEMINI_API_KEY)

# Initialisation du modèle Gemini avec instruction système
# Utilisation du modèle : gemini-2.5-flash-lite
model_gemini = genai.GenerativeModel(
    'gemini-2.5-flash-lite',
    system_instruction="Tu dois TOUJOURS commencer ta réponse par cette phrase exacte : 'En tant qu'analyste macro-économique spécialisé dans le Forex, voici mon analyse et mes recommandations concernant cette annonce :'"
)

def call_llm(prompt):
    """
    Appelle l'API Gemini pour générer une analyse.

    Args:
        prompt: Le prompt formaté avec le contexte DistilRoBERTa

    Returns:
        str: La réponse générée par Gemini
    """
    try:
        response = model_gemini.generate_content(
            prompt,
            generation_config=genai.types.GenerationConfig(
                temperature=0.7,
                max_output_tokens=2000,
            )
        )
        return response.text
    except Exception as e:
        return f"Erreur lors de l'appel à Gemini : {str(e)}"

# Test de l'appel au LLM
print("=== ANALYSE GÉNÉRÉE PAR GEMINI (gemini-2.5-flash-lite) ===\n")
# Note: Assurez-vous d'avoir exécuté les cellules précédentes pour avoir la variable 'prompt'
if 'prompt' in locals():
    analysis = call_llm(prompt)
    print(analysis)
else:
    print("Erreur : La variable 'prompt' n'est pas définie. Veuillez exécuter l'étape précédente d'abord.")

=== ANALYSE GÉNÉRÉE PAR GEMINI (gemini-2.5-flash-lite) ===

En tant qu'analyste macro-économique spécialisé dans le Forex, voici mon analyse et mes recommandations concernant cette annonce :

**1. EXPLICATION :**
L'annonce concerne le taux de chômage au Canada. Ce chiffre mesure le pourcentage de la population active qui est sans emploi et à la recherche active d'un emploi. Un taux de chômage plus bas est généralement considéré comme un signe de santé économique, car il indique que plus de personnes travaillent, ce qui stimule la consommation et la croissance.

**2. ANALYSE DE LA SURPRISE :**
Le taux de chômage annoncé (6.9%) est inférieur aux prévisions (7.1%). C'est une "bonne surprise" pour le dollar canadien (CAD). Une réduction du chômage par rapport aux attentes suggère que l'économie canadienne est plus robuste que prévu, ce qui peut attirer les investisseurs et augmenter la demande pour la devise.

**3. VALIDATION DU SENTIMENT :**
La prédiction du modèle d'analyse de sentiment 

## 14. Fonction Pipeline Complète (One-Shot)

In [19]:
def analyze_forex_news_v2(news_text, reference_date=None, top_k=5, verbose=True):
    """
    🚀 PIPELINE COMPLET V2 - Assistant Macro-Économique Forex

    Combine :
    1. DistilRoBERTa Embeddings (recherche de similarité)
    2. DistilRoBERTa Classifier (sentiment 3 classes)
    3. Gemini LLM (génération d'analyse)

    Args:
        news_text: Annonce au format Embedding_Context
                   Ex: "USD CPI m/m. Actual: 0.4%, Forecast: 0.2%. Usual Effect: ..."
        reference_date: Date pour filtrer les précédents (format 'YYYY-MM-DD')
        top_k: Nombre de précédents historiques à récupérer
        verbose: Afficher les détails intermédiaires

    Returns:
        dict: Résultats complets (similar_events, sentiment, analysis)
    """
    if verbose:
        print("=" * 70)
        print("🚀 ASSISTANT MACRO-ÉCONOMIQUE FOREX V2 (DistilRoBERTa + Gemini)")
        print("=" * 70)
        print(f"\n📝 Annonce : {news_text[:100]}...")

    # ÉTAPE 1 : Recherche de précédents
    if verbose:
        print("\n📊 Étape 1 : Recherche de précédents historiques...")
    similar_events = find_similar_events_v2(news_text, data, top_k=top_k, query_date=reference_date)

    # ÉTAPE 2 : Analyse de sentiment
    if verbose:
        print("\n🧠 Étape 2 : Analyse de sentiment (3 classes)...")
    sentiment = get_sentiment_distilroberta(news_text)
    if verbose:
        print(f"   → Prédiction : {sentiment['class_name']} ({sentiment['score']:.1%})")

    # ÉTAPE 3 : Génération du prompt
    if verbose:
        print("\n📝 Étape 3 : Génération du prompt...")
    prompt = generate_analysis_v2(news_text, similar_events, sentiment)

    # ÉTAPE 4 : Appel à Gemini
    if verbose:
        print("\n🤖 Étape 4 : Appel à Gemini...")
    analysis = call_gemini(prompt)

    if verbose:
        print("\n" + "=" * 70)
        print("📋 ANALYSE FINALE")
        print("=" * 70)
        print(analysis)

    return {
        'news': news_text,
        'similar_events': similar_events,
        'sentiment': sentiment,
        'prompt': prompt,
        'analysis': analysis
    }

print("✅ Fonction analyze_forex_news_v2() prête")
print("   Usage : result = analyze_forex_news_v2('USD CPI m/m. Actual: 0.4%...')")

✅ Fonction analyze_forex_news_v2() prête
   Usage : result = analyze_forex_news_v2('USD CPI m/m. Actual: 0.4%...')


In [20]:
Final_test = analyze_forex_news_v2('EUR Unemployment Rate. Actual: 7.3%, Forecast: 6.6%. Usual Effect: Actual less than Forecast is good for currency.')

🚀 ASSISTANT MACRO-ÉCONOMIQUE FOREX V2 (DistilRoBERTa + Gemini)

📝 Annonce : EUR Unemployment Rate. Actual: 7.3%, Forecast: 6.6%. Usual Effect: Actual less than Forecast is good...

📊 Étape 1 : Recherche de précédents historiques...
📅 Filtrage : 15490 événements avant 2025-12-13


Extraction embeddings: 100%|██████████| 1/1 [00:00<00:00,  2.81it/s]



🧠 Étape 2 : Analyse de sentiment (3 classes)...
   → Prédiction : NEUTRE (41.4%)

📝 Étape 3 : Génération du prompt...

🤖 Étape 4 : Appel à Gemini...

📋 ANALYSE FINALE
En tant qu'analyste macro-économique spécialisé dans le Forex, voici mon analyse et mes recommandations concernant cette annonce :

**1. EXPLICATION :**
L'annonce concerne le taux de chômage dans la zone euro. Il s'agit du pourcentage de la population active qui est sans emploi mais recherche activement un travail. Un taux de chômage élevé indique une faiblesse de l'économie, tandis qu'un taux bas suggère une économie en bonne santé.

**2. ANALYSE DE LA SURPRISE :**
Le taux de chômage annoncé est de 7.3%, ce qui est significativement plus élevé que les prévisions qui étaient de 6.6%. Selon la règle générale, un taux de chômage *inférieur* aux prévisions est bon pour la devise. Dans ce cas, le taux réel est *supérieur* aux prévisions, ce qui constitue une **mauvaise surprise** pour l'euro.

**3. VALIDATION DU SENTIMENT :*

======================================================================
🚀 ASSISTANT MACRO-ÉCONOMIQUE FOREX V2 (DistilRoBERTa + Gemini)
======================================================================

📝 Annonce : EUR Unemployment Rate. Actual: 7.3%, Forecast: 6.6%. Usual Effect: Actual less than Forecast is good...

📊 Étape 1 : Recherche de précédents historiques...
📅 Filtrage : 15490 événements avant 2025-12-13
Extraction embeddings: 100%|██████████| 1/1 [00:00<00:00,  2.81it/s]

🧠 Étape 2 : Analyse de sentiment (3 classes)...
   → Prédiction : NEUTRE (41.4%)

📝 Étape 3 : Génération du prompt...

🤖 Étape 4 : Appel à Gemini...

======================================================================
📋 ANALYSE FINALE
======================================================================
En tant qu'analyste macro-économique spécialisé dans le Forex, voici mon analyse et mes recommandations concernant cette annonce :

**1. EXPLICATION :**
L'annonce concerne le taux de chômage dans la zone euro. Il s'agit du pourcentage de la population active qui est sans emploi mais recherche activement un travail. Un taux de chômage élevé indique une faiblesse de l'économie, tandis qu'un taux bas suggère une économie en bonne santé.

**2. ANALYSE DE LA SURPRISE :**
Le taux de chômage annoncé est de 7.3%, ce qui est significativement plus élevé que les prévisions qui étaient de 6.6%. Selon la règle générale, un taux de chômage *inférieur* aux prévisions est bon pour la devise. Dans ce cas, le taux réel est *supérieur* aux prévisions, ce qui constitue une **mauvaise surprise** pour l'euro.

**3. VALIDATION DU SENTIMENT :**
Le modèle d'analyse de sentiment prédit un résultat "NEUTRE" avec une confiance de 41.4%. Cette prédiction est quelque peu en contradiction avec la mauvaise surprise des chiffres annoncés. Bien que la probabilité de mouvement Neutre soit la plus élevée, la divergence entre le chiffre réel et le chiffre prévu suggère un potentiel de mouvement. La faible confiance du modèle (41.4%) renforce l'idée que le marché pourrait réagir.

**4. ANALYSE HISTORIQUE :**
Les précédents historiques montrent des réactions mitigées mais avec une tendance à la neutralité pour des annonces similaires.
*   Le cas le plus similaire (score 0.9998) en date du 30 octobre 2007 avec un taux réel de 7.3% (supérieur au forecast de 6.9%) a entraîné une hausse de la devise (0.1603%). Ceci est une anomalie par rapport à la théorie classique.
*   Les autres précédents très similaires (scores > 0.9996) montrent des réactions neutres (0.0000% de variation), même lorsque le taux réel était égal ou légèrement supérieur aux prévisions.

Le pattern observé est donc une tendance à la neutralité du marché, malgré des divergences entre les chiffres réels et prévus, avec une seule exception notable où une mauvaise surprise a curieusement mené à une hausse.

**5. SCÉNARIOS PROBABLES :**

*   **Scénario Haussier (pour l'Euro) :**
    *   **Conditions :** Le marché ignore la mauvaise surprise des chiffres du chômage et se concentre sur d'autres facteurs macroéconomiques positifs pour l'euro, ou réagit de manière irrationnelle comme observé dans un précédent historique. Une intervention verbale des banquiers centraux pour rassurer pourrait aussi soutenir l'euro.
    *   **Probabilité :** Faible (environ 27.5% selon l'analyse de sentiment, potentiellement plus basse si le marché suit la logique économique).

*   **Scénario Baissier (pour l'Euro) :**
    *   **Conditions :** Le marché réagit négativement à la mauvaise surprise, interprétant le taux de chômage élevé comme un signe de faiblesse économique et un potentiel frein pour la BCE dans son resserrement monétaire. La déviation significative par rapport aux prévisions renforce ce scénario.
    *   **Probabilité :** Modérée (environ 31.1% selon l'analyse de sentiment, potentiellement plus élevée étant donné la mauvaise surprise).

**6. PAIRES À SURVEILLER :**
*   **EUR/USD :** La paire la plus liquide et la plus directement impactée par les données de la zone euro.
*   **EUR/GBP :** Les divergences entre la politique monétaire de la BCE et de la Banque d'Angleterre pourraient être exacerbées par cette annonce.
*   **EUR/JPY :** Le yen japonais, souvent considéré comme une valeur refuge, pourrait réagir à une faiblesse de l'euro.

**7. RECOMMANDATION :**
La situation est ambiguë. La mauvaise surprise des chiffres du chômage devrait logiquement peser sur l'euro, mais les précédents historiques et l'analyse de sentiment suggèrent une réaction potentiellement neutre, voire surprenante.

*   **Prudence accrue :** En raison de cette ambiguïté, il est conseillé de faire preuve d'une prudence extrême.
*   **Attendre confirmation :** Ne prenez pas de position immédiate basée uniquement sur cette annonce. Attendez de voir comment le marché réagit dans les heures qui suivent la publication.
*   **Surveiller les autres indicateurs :** L'euro est influencé par de nombreux facteurs. Surveillez d'autres annonces économiques importantes pour la zone euro et les États-Unis (pour EUR/USD) dans les jours à venir.
*   **Considérer des stratégies de trading moins directionnelles :** Si vous souhaitez trader, envisagez des stratégies qui bénéficient de la volatilité mais sont moins sensibles à la direction exacte du marché, ou attendez une confirmation claire du mouvement.
*   **Gestion du risque :** Si vous décidez de prendre position, utilisez des stops loss serrés pour limiter les pertes potentielles en cas de mouvement inattendu.

En résumé, bien que la statistique soit négative pour l'euro, le marché pourrait ne pas réagir fortement, ou même réagir à l'inverse. La meilleure approche est d'observer le comportement du marché après la publication avant de prendre une décision.